File: Chain-of-Thought (CoT) Pattern/langgraph_python_example/cot_agent.py
# Chain-of-Thought (CoT) Pattern - LangGraph Python Example

This example demonstrates the Chain-of-Thought Pattern using LangGraph and Python. The agent receives a problem, prompts the LLM to reason step by step, and returns the reasoning and answer.

## Requirements

- Python 3.8+
- `langgraph` library
- `python-dotenv` (for .env support)
- Mistral LLM API access

## Install dependencies

```bash
pip install langgraph python-dotenv requests
```

## Example Code

```python
import os
from langgraph import Agent, Environment, LLM
from dotenv import load_dotenv

load_dotenv()

MISTRAL_API_KEY = os.getenv("MISTRAL_API_KEY")
MISTRAL_API_URL = "https://api.mistral.ai/v1/chat/completions"

class SimpleEnvironment(Environment):
    def get_observation(self):
        return input("Problem: ")
    def send_action(self, action):
        print(f"Agent reasoning:\n{action}")

class CoTAgent(Agent):
    def __init__(self, llm):
        self.llm = llm
    def act(self, observation):
        prompt = f"Let's solve this step by step. {observation}\nExplain your reasoning at each step, then give the final answer."
        return self.llm.complete(prompt)

llm = LLM(
    provider="mistral",
    api_key=MISTRAL_API_KEY,
    api_url=MISTRAL_API_URL,
)

env = SimpleEnvironment()
agent = CoTAgent(llm)

observation = env.get_observation()
action = agent.act(observation)
env.send_action(action)
```

---

- Try a math or logic problem to see step-by-step reasoning.
- Make sure your `.env` file contains your Mistral API key.

File: Chain-of-Thought (CoT) Pattern/autogen_typescript_example/cot_agent.ts
// Chain-of-Thought (CoT) Pattern - Autogen TypeScript Example
// To run: npm install && npm run cot-agent

import axios from 'axios';
import * as readline from 'readline';
import * as dotenv from 'dotenv';
dotenv.config();

const MISTRAL_API_URL = 'https://api.mistral.ai/v1/chat/completions';
const MISTRAL_API_KEY = process.env.MISTRAL_API_KEY;

function buildCoTPrompt(userInput: string): string {
  return `Let's solve this step by step. ${userInput}\nExplain your reasoning at each step, then give the final answer.`;
}

async function cotAgent(userInput: string): Promise<string> {
  const prompt = buildCoTPrompt(userInput);
  const response = await axios.post(
    MISTRAL_API_URL,
    {
      model: 'mistral-tiny',
      messages: [{ role: 'user', content: prompt }],
    },
    {
      headers: {
        'Authorization': `Bearer ${MISTRAL_API_KEY}`,
        'Content-Type': 'application/json',
      },
    }
  );
  return response.data.choices[0].message.content;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('Problem: ', async (userInput: string) => {
  try {
    const agentResponse = await cotAgent(userInput);
    console.log('Agent reasoning:\n', agentResponse);
  } catch (err) {
    console.error('Error:', err);
  }
  rl.close();
});
